nixos/modules/services/web-apps/plausible.nix

```diff
@@ -0,0 +1,270 @@
+          postgres (`services.plausible.database.postgres.setup') is enabled!
+        '';
+      }
+    ];
+
+    services.postgresql = mkIf cfg.database.postgres.setup {
+      enable = true;
+    };
+
+    services.clickhouse = mkIf cfg.database.clickhouse.setup {
+      enable = true;
+    };
+
+    systemd.services = mkMerge [
+      {
+        plausible = {
+          inherit (pkgs.plausible.meta) description;
+          documentation = [ "https://plausible.io/docs/self-hosting" ];
+          wantedBy = [ "multi-user.target" ]
+            ++ optional cfg.database.clickhouse.setup "clickhouse.service"
+            ++ optional cfg.database.postgres.setup "postgresql.service";
+          after = optional cfg.database.postgres.setup "plausible-postgres.service";
+
+          environment = {
+            # NixOS specific option to avoid that it's trying to write into its store-path.
+            # See also https://github.com/lau/tzdata#data-directory-and-releases
+            PLAUSIBLE_TZDATA = "/var/lib/plausible/elixir_tzdata";
+
+            # Configuration options from
+            # https://plausible.io/docs/self-hosting-configuration
+            PORT = toString cfg.server.port;
+            DISABLE_REGISTRATION = boolToString cfg.server.disableRegistration;
+
+            ADMIN_USER_NAME = cfg.adminUser.name;
+            ADMIN_USER_EMAIL = cfg.adminUser.email;
+
+            DATABASE_SOCKET_DIR = cfg.database.postgres.socket;
+            DATABASE_NAME = cfg.database.postgres.dbname;
+            CLICKHOUSE_DATABASE_URL = cfg.database.clickhouse.url;
+
+            BASE_URL = cfg.server.baseUrl;
+
+            MAILER_EMAIL = cfg.mail.email;
+            SMTP_HOST_ADDR = cfg.mail.smtp.hostAddr;
+            SMTP_HOST_PORT = toString cfg.mail.smtp.hostPort;
+            SMTP_RETRIES = toString cfg.mail.smtp.retries;
+            SMTP_HOST_SSL_ENABLED = boolToString cfg.mail.smtp.enableSSL;
+            ${if cfg.mail.smtp.user != null then "SMTP_USER_NAME" else null} = cfg.mail.smtp.user;
+          };
+
+          path = [ pkgs.plausible ]
+            ++ optional cfg.database.postgres.setup config.services.postgresql.package;
+
+          serviceConfig = {
+            DynamicUser = true;
+            PrivateTmp = true;
+            WorkingDirectory = "/var/lib/plausible";
+            StateDirectory = "plausible";
+            ExecStartPre = "@${pkgs.writeShellScript "plausible-setup" ''
+              ${envSecrets}
+              ${pkgs.plausible}/createdb.sh
+              ${pkgs.plausible}/migrate.sh
+              ${optionalString cfg.adminUser.activate ''
+                if ! ${pkgs.plausible}/init-admin.sh | grep 'already exists'; then
+                  psql -d plausible <<< "UPDATE users SET email_verified=true;"
+                fi
+              ''}
+            ''} plausible-setup";
+            ExecStart = "@${pkgs.writeShellScript "plausible" ''
```

Comment:
Almost enough to make you want to simplify and use `EnvironmentFile` to back each of the secrets options...

Reply:
IIRC `EnvironmentFile` doesn't allow file-substitutions such as `$(<...)` and I thought that having `xxxFile`-options is generally nicer because it can be trivially integrated with e.g. `sops-nix` or NixOps' `deployment.keys` feature.

Reply:
I've noticed the pattern that people ask for both the key and value in secrets files.

You still could have an option per file, but simply append each option to the `EnvironmentFile` list.

I'm mostly just mentioning this as an alternative to avoid shell scripting. Both ways have pros and cons.

Reply:
Just as an additional data point, I use EnvironmentFile with agenix (incidentally on an elixir service), and the integration works really well.

On a separate note, systemd doesn't recommend environment variable or EnvironmentFile for secrets. I've been trying to make LoadCredentials work (what they actually recommend) without success.

Reply:
I'm aware they don't recommend `EnvironmentFile` for secrets... but they don't have a world readable nix store to deal with 😄 

You're right, though. I should have mentioned `LoadCredential`, especially because this runs as `DynamicUser`, which implies these secrets are already world readable the way this module is.

Great suggestion!

Reply:
There is an ongoing conversation regarding LoadCredential at the moment with the systemd guys.
It is broken with the current systemd (247) but it should be fixed with the next one. There is a PR for it at the moment but not merged.
Second point is you can only use files for credentials. I'm still a bit confused how this should be used with environment variable for example. Should you separate all your env vars into separate files and then source them in your web application? I'm not sure.
Trying to give more information in case you are considering LoadCredential. Not making a recommendation for it though.

Reply:
>  Second point is you can only use files for credentials. I'm still a bit confused how this should be used with environment variable for example. Should you separate all your env vars into separate files and then source them in your web application? I'm not sure.

Perhaps something like `export FOO=$(<secret)`? Basically what I'm doing now except that you have to manage these files somehow else (e.g. with sops-nix or `deployment.keys`).

Reply:
Thanks for the recommendation, I'm asking them directly, will let you know if I find out anything.